lib/checksizeof.cpp

```diff
@@ -317,6 +317,46 @@ void CheckSizeof::sizeofCalculationError(const Token *tok, bool inconclusive)
                 "sizeofCalculation", "Found calculation inside sizeof().", CWE682, inconclusive);
 }
 
+//-----------------------------------------------------------------------------
+
+void CheckSizeof::sizeofFunction()
+{
+    if (!_settings->isEnabled(Settings::WARNING))
+        return;
+
+    for (const Token *tok = _tokenizer->tokens(); tok; tok = tok->next()) {
+        if (Token::simpleMatch(tok, "sizeof (")) {
+
+            // ignore if the `sizeof` result is cast to void inside a macro, i.e. the calculation is
+            // expected to be parsed but skipped, such as in a disabled custom ASSERT() macro
+            if (tok->isExpandedMacro() && tok->previous()) {
+                const Token *cast_end = (tok->previous()->str() == "(") ? tok->previous() : tok;
+                if (Token::simpleMatch(cast_end->tokAt(-3), "( void )") ||
+                    Token::simpleMatch(cast_end->previous(), "static_cast<void>")) {
+                    continue;
+                }
+            }
+
+            for (const Token *argument = tok->next()->astOperand2(); argument; argument = argument->astOperand2()) {
+                const Token *checkToken = argument->previous();
```

Comment:
why do you have a loop, is there a test case where it is not enough to only look at the first token?

Reply:
Yes, when doing `sizeof(x.foo())`, but let me double check.

Reply:
You can rewrite that as ’sizeof((x.foo)())’. Then you can intuitively see that the operands for ’.’ are x and foo.

Reply:
You are right, I was doing something else wrong. I updated to just check the first token.